## Overview
In this exercise we will:
    - load some data
    - inspect it
    - do some cleaning
    - model our data
    - load it into the graph
    - gain some insights
    
This exercise is left largely open and the commented cells are only intended as a guideline. Please try to implement some solution yourself before looking at the solution.

In [1]:
import pandas as pd
import neomodel

In [2]:
# Configure Neo4J

from neomodel import config

user = "neo4j"
password = "password"

config.DATABASE_URL = f'bolt://{user}:{password}@localhost:7687'
    
def drop_delete_everything():
    '''Helpful to wipe the db properly'''
    db.cypher_query("MATCH (n) DETACH DELETE n")
    
    neomodel.core.remove_all_labels()

# Now that we have things setup lets look at our data

It's a simple collection records of letters sent between several people. You can assume that there is only one person with each name. i.e. it is the same Will in all entries marked Will

In [4]:
data = pd.read_csv("./letters.csv", names=["Sender", "Reciever", "date"])
data

Sender          Reciever               date
0                 John              Will          25.5.1801
1                Scott              Will          13.3.1805
2                Sarah             Kevin         24.11.1807
3              Michael              Will         24.10.1803
4              Michael              Will           5.2.1805
5              Michael  Wordsworth, John          17.2.1805
6                 Lisa             Kevin               1805
7                 Lisa             Kevin               1805
8              Michael              Will           3.3.1805
9              Michael              Will          24.3.1805
10             Michael              Will          21.7.1805
11             Michael              Will          11.8.1805
12             Michael              Will         25.11.1805
13             Michael              Will           5.3.1806
14             Michael              Will          29.6.1806
15             Michael              Will           6.8.1806
16             Michael              Will          10.8.1806
17             Michael              Will               1806
18             Michael              Will          11.9.1806
19             Michael              Will         14.10.1806
20             Michael              Will          6.11.1806
21             Michael              Will          15.3.1807
22             Michael              Will          25.2.1808
23                Lisa             Kevin           1.5.1808
24             Michael              Will               1808
25                Lisa             Kevin         12.12.1808
26             Caoimhe              Will          14.3.1805
27               Sarah              Rory          10.9.1799
28               Sarah              Will  23.7.1803/ 8.1803
29               Sarah              Will           8.2.1804
..                 ...               ...                ...
177     Marshall, Jane             Kevin          24.2.1805
178     Marshall, John              Will           1.9.1807
179              Basil              Will          16.4.1804
180             Robert              Will               1805
181              Tobin         21.2.1805                NaN
182              Tobin         14.4.1805                NaN
183  Wilkinson, Thomas              Will          29.7.1806
184       Wilson, John              Will          24.5.1802
185               Will             Kevin                NaN
186               Will             Kevin                NaN
187               Will             Kevin          26.3.1808
188      Graham, James              Rory          6.10.1802
189               Will              Rory          24.6.1802
190               Will              Rory           3.7.1802
191               Rory              Will           9.7.1802
192               Rory              Will          14.2.1805
193         Barton, T.              Rory          2.11.1804
194         Barton, T.              Rory          7.12.1804
195        Myers, John              Rory          27.5.1807
196     Cooper, Thomas              Rory         11.10.1800
197     Smith, Matthew              Rory          29.1.1804
198   Jennings, Thomas              Rory          7.12.1806
199     Hutton, Thomas              Rory          11.3.1806
200              Basil              Rory          12.6.1802
201      Stewart, C.H.              Rory          14.5.1801
202               Will              Rory          18.6.1802
203        Cookson, W.              Rory          22.6.1802
204   Wordsworth, Mr/s              Rory          25.5.1802
205              Tobin              Rory          25.4.1807
206              Basil              Rory          19.6.1802

[207 rows x 3 columns]

## Is it clean? (hint: it's not)
## How can I account for the missing values/ wrong types

In [ ]:
# Do some cleaning here





# Construct the data model

In [ ]:
# Data model






In [5]:
# Load the data into the graph







How many nodes are in the graph? - Answer: 44

How many connections are in the graph? - Answer: 207

Which node is most connected? - Answer: Rory